### DISCUSSION TITLE
Identifying a "special" comment

### DISCUSSION THREAD

**Discussion by tinygiant98** [tree-sitter/tree-sitter#1610] [Q&A] [answered]
I've built a tree-sitter parser for a custom language that's based generally on C, so I stole tree-sitter-c grammar and heavily modified it.  It works well for just about every situation I've run into.  I've just received a request for a new feature that identifies annotation/decorators within comments, something like:
```c
// @ name [data:more data]
```
I've created a rule that identifies the annotation portion (the part after the `//`), which the parser identifies just fine when it's not part of a comment; However, anytime it's part of a comment, the parser identifies it as a comment.  I've tried setting the rule as a top-level-item, using precedence, and making it a part of the comment rule choice options, but cannot seem to get this method to work.  Is it possible to identify an annotation within a comment as a special type versus a simple comment?

**Comment by ahelwer** [comment, marked as answer]
I'm assuming what you mean is that your comment regex just consumes everything to the end of the line instead of tokenizing it like you want. This is a somewhat tricky lexical precedence issue that could possibly be solved by assigning lexical precedence inside the token function. Note that lexical precedence is distinct from parsing precedence.

**Reply by tinygiant98** [reply]
Ok, thanks, I'll start looking at that and see what I can figure out.

**Reply by tinygiant98** [reply]
So, now that I look closer at it, since I'm trying to return a specific node type for the custom decoration, I really just want to have the decoration rule return at a higher precedence than the comment rule, but it seems to be setting the comment at a higher precedent when determining what rule is satisfied.  If both the decoration rule and comments are considered top_level_items, how do I tell the grammar that I would rather check for decoration first, then comment?

**Reply by dfgordon** [reply]
Expanding on what @ahelwer said, if a rule is pure regex, I find the lexer will consume that greedily and the parser will commit to it very strongly no matter what other rules or precedence settings are in the grammar.  In other words `comment: $ => RegExp('//.*')` will tend to suppress any alternatives, whereas `comment: $ => seq('//',/.*/)` will allow for them.

**Comment by tinygiant98** [comment]
@dfgordon  Since the comment rules are all `option(seq)` rules, I used your comment to modify the decorator into a pure regular expression match instead of a seq match and it worked.  Thanks for the info.

**Reply by dfgordon** [reply]
OK interesting, I guess you can also go the opposite route.  Actually I think my second rule example was not great, `/.*/` is probably bad, maybe should have been `repeat($.anychar)` or something.
